# Date Night Movie

#### Grading:


- Code: 90 pts
- Markdown Documentation: 10 pts


In this assignment we are going to use pandas to figure out - What's the best **date-night movie**?

This assignment is going to use
- Joining
- Groupby
- Sorting


In [1]:
import os
import pandas as pd

##### Read in the movie data: `pd.read_table`

In [2]:
def get_movie_data():
    
    unames = ['user_id','gender','age','occupation','zip']
    users = pd.read_table(os.path.join('../data','users.dat'), 
                          sep='::', header=None, names=unames)
    
    rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
    ratings = pd.read_table(os.path.join('../data', 'ratings.dat'), 
                            sep='::', header=None, names=rnames)
    
    mnames = ['movie_id', 'title','genres']
    movies = pd.read_table(os.path.join('../data', 'movies.dat'), 
                           sep='::', header=None, names=mnames)

    return users, ratings, movies

In [3]:
users, ratings, movies = get_movie_data()

/home/rashmi/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """
/home/rashmi/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:9: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  if __name__ == '__main__':
/home/rashmi/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:13: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  del sys.path[0]


In [4]:
users.head()

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455

In [5]:
ratings.head()

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

In [6]:
movies.head()

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

##### Clean up the `movies`

- Get the `year`
- Shorten the `title`


In [7]:
tmp = movies.title.str.extract('(.*) \(([0-9]+)\)')
tmp.apply(lambda x:x[0] if len(x) > 0 else None)
tmp.apply(lambda x: x[0][:40] if len(x) > 0 else None)
print(tmp)

                                0     1
0                       Toy Story  1995
1                         Jumanji  1995
2                Grumpier Old Men  1995
3               Waiting to Exhale  1995
4     Father of the Bride Part II  1995
...                           ...   ...
3878             Meet the Parents  2000
3879          Requiem for a Dream  2000
3880                    Tigerland  2000
3881             Two Family House  2000
3882               Contender, The  2000

[3883 rows x 2 columns]


In [8]:
movies['year'] = tmp[1]
movies['short_title'] = tmp[0]

In [9]:
movies.head()

movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II

In [10]:
movies.head()

movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II

##### Join the tables with `pd.merge` (20 pts)

#### First merge users and rating table using inner join on user_id

In [11]:
userRatings = pd.merge(users,ratings)
userRatings.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp
0        1      F    1          10  48067      1193       5  978300760
1        1      F    1          10  48067       661       3  978302109
2        1      F    1          10  48067       914       3  978301968
3        1      F    1          10  48067      3408       4  978300275
4        1      F    1          10  48067      2355       5  978824291

#### Merge the userRating table movies table using inner join on movie_id

In [12]:
userMovieRatings = pd.merge(userRatings,movies)
userMovieRatings.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  year  \
0  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
1  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
2  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
3  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
4  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   

                       short_title  
0  One Flew Over the Cuckoo's Nest  
1  One Flew Over the Cuckoo's Nest  
2  One Flew Over the Cuckoo's Nest  
3  One Flew Over the Cuckoo's Nest  
4  One Flew Over the Cuckoo's Nest

In [13]:
#Export the merged table to csv 
userMovieRatings.to_csv('../data/userMovieRatingsMergedData.csv',index=False)

#### What's the highest rated movie? (20 pts))

#### Method 1
- Group by movie_id 
- Calculate the count of people who have rated the movie: This considers the number of people who have rated the movie
- Calculate the mean of the ratings  
- Sort the records by count and mean in descending order
- Select the movie with the heighest number of votes and with highest mean rating

In [14]:
#Groupby movie_id and caluculate the count of people who have rated the movie and the mean of the ratings and sort first by count and then by mean in descending order
highRatedMovie = userMovieRatings.groupby('movie_id')['rating'].agg(['count','mean']).reset_index().sort_values(['count','mean'],ascending=[False,False])
highRatedMovie

movie_id  count      mean
2651      2858   3428  4.317386
253        260   2991  4.453694
1106      1196   2990  4.292977
1120      1210   2883  4.022893
466        480   2672  3.763847
...        ...    ...       ...
3013      3237      1  1.000000
3086      3312      1  1.000000
3146      3376      1  1.000000
3227      3460      1  1.000000
3658      3904      1  1.000000

[3706 rows x 3 columns]

In [15]:
# Get the movie details of the first high rated movie
movies[movies['movie_id'] == highRatedMovie['movie_id'].iloc[0]]

movie_id                   title        genres  year      short_title
2789      2858  American Beauty (1999)  Comedy|Drama  1999  American Beauty

#### American Beauty released in 1999 with movie_id =2858 is the highest rated movie

#### Method 2
- Group by movie_id 
- Calculate the sum of the ratings of the movie : This considers the number of people who have rated the movie and also their rating
- Calculate the mean and meadian of the ratings
- Sort the records by sum and mean in descending order
- Select the movie with the height sum of rating and heighest mean rating

In [16]:
#group the merged data by movie_id
highRatedMovie_m2 = userMovieRatings.groupby('movie_id')
#Calculate the sum mean and meadian and sort it by sum and mean in descending order
highRatedMovie_m2 = highRatedMovie_m2['rating'].agg(['sum','median','mean']).reset_index().sort_values(['sum','mean'],ascending=[False,False])
highRatedMovie_m2

movie_id    sum  median      mean
2651      2858  14800     5.0  4.317386
253        260  13321     5.0  4.453694
1106      1196  12836     4.0  4.292977
1120      1210  11598     4.0  4.022893
1848      2028  11507     5.0  4.337354
...        ...    ...     ...       ...
3013      3237      1     1.0  1.000000
3086      3312      1     1.0  1.000000
3146      3376      1     1.0  1.000000
3227      3460      1     1.0  1.000000
3658      3904      1     1.0  1.000000

[3706 rows x 4 columns]

In [17]:
# Get the movie details of the first high rated movie
movies[movies['movie_id'] == highRatedMovie_m2['movie_id'].iloc[0]]

movie_id                   title        genres  year      short_title
2789      2858  American Beauty (1999)  Comedy|Drama  1999  American Beauty

#### American Beauty released in 1999 with movie_id =2858 is the highest rated movie

In [18]:
#Display the movie_id's of top 5 rated movies and their average rating
highRatedMovie_m2.head(5)

movie_id    sum  median      mean
2651      2858  14800     5.0  4.317386
253        260  13321     5.0  4.453694
1106      1196  12836     4.0  4.292977
1120      1210  11598     4.0  4.022893
1848      2028  11507     5.0  4.337354

#### Note
- When there is tie in the sum of ratings , we are resolving the tie by giving preference to the rating as we are sorting by sum first and then mean.



###### What is a good rated movie for date night? (60 pts)

- Hint - highly rated movie by 
    - both partners (might be the same gender or not),
    - based on genre preferences,
    - age group can also be combined

### Preferences for watching a movie on a date night
- We would like to watch a romantic comedy(romcom) movie
- Movie that is mostly prefered(rated) by the people in the age of 25 or 30
- Movie which is released in the year 1997-2000
- Move which has highest ratings with the above mentioned preferences


In [19]:
#Filter the data based on the year between 1997 and 2000
filterdata = userMovieRatings[userMovieRatings['year'].between('1997','2000',inclusive=True)]
#Filter the data based on genere as Comedy/Romance 
filterdata = filterdata[filterdata['genres'] == 'Comedy|Romance']
#Filter the data based on the pople with age 25 and 30
filterdata = filterdata[filterdata['age'].isin([25,30])]
#Calculate the highest rated movie
highRatedRomCom = filterdata.groupby('movie_id')['rating'].agg(['sum','mean']).reset_index().sort_values(['sum','mean'],ascending=[False,False])
highRatedRomCom.head()

movie_id   sum      mean
22      2396  3535  4.081986
28      2502  1474  3.760204
11      1777  1456  3.559902
7       1569  1095  3.348624
35      2671  1047  3.549153

In [20]:
#Movie Details of the top rated movie which matches all of our preferences
movies[movies['movie_id']==highRatedRomCom['movie_id'].iloc[0] ]

movie_id                       title          genres  year  \
2327      2396  Shakespeare in Love (1998)  Comedy|Romance  1998   

              short_title  
2327  Shakespeare in Love

In [21]:
#Details of the date night movie chosen is displayed as it matches all the preferences
userMovieRatings[(userMovieRatings['short_title']=='Shakespeare in Love') & (userMovieRatings['age'].isin([25,30]))]

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
144107        8      M   25          12  11413      2396       5  978229524   
144109       11      F   25           1  04093      2396       2  978902561   
144111       15      M   25           7  22903      2396       4  978196817   
144115       24      F   25           7  10023      2396       4  978134987   
144116       28      F   25           1  14607      2396       5  978125846   
...         ...    ...  ...         ...    ...       ...     ...        ...   
146467     6024      M   25          12  53705      2396       4  956749422   
146468     6025      F   25           1  32607      2396       4  956731134   
146469     6035      F   25           1  78734      2396       4  956712860   
146470     6036      F   25          15  32603      2396       2  956710191   
146473     6040      M   25           6  11106      2396       3  956704475   

                             title          genres  year          short_title  
144107  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  
144109  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  
144111  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  
144115  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  
144116  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  
...                            ...             ...   ...                  ...  
146467  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  
146468  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  
146469  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  
146470  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  
146473  Shakespeare in Love (1998)  Comedy|Romance  1998  Shakespeare in Love  

[866 rows x 12 columns]

### Date Night Movie choosen is "Shakespeare in Love"
- It is a RomCom produced in the year 1998 and is higly rated by the people with age 25 or 30, so it satisfies all the preferences